src/fields/Fields.cpp

```diff
@@ -166,33 +164,79 @@ struct interpolated_field_inner {
-                                             j_cell+iy, lo2));
+                field_value += sx_cell[ix] * sy_cell[iy] * array(i_cell+ix, j_cell+iy, args...);
             }
         }
         return field_value;
     }
 };
 
-template<int interp_order_xy>
-struct interpolated_field {
+template<int interp_order_xy, class MfabType>
+struct interpolated_field_xy {
     // use brace initialization as constructor
-    amrex::MultiFab f_view_this; // field to interpolate on this slice
-    amrex::MultiFab f_view_prev; // field to interpolate on previous slice
-    const amrex::Geometry& geom; // geometry of field
+    MfabType mfab; // MultiFab type object of the field
+    amrex::Geometry geom; // geometry of field
+
+    // use .array(mfi) like with amrex::MultiFab
+    auto array (amrex::MFIter& mfi) const {
+        auto mfab_array = mfab.array(mfi);
+        return interpolated_field_xy_inner<interp_order_xy, decltype(mfab_array)>{
+            mfab_array, 1/geom.CellSize(0), 1/geom.CellSize(1),
+            GetPosOffset(0, geom, geom.Domain()), GetPosOffset(1, geom, geom.Domain())};
+    }
+};
+
+struct interpolated_field_z_inner {
+    // captured variables for GPU
+    amrex::Array4<amrex::Real const> arr_this;
+    amrex::Array4<amrex::Real const> arr_prev;
+    amrex::Real rel_z;
+    int lo2;
+
+    // linear longitudinal field interpolation
+    AMREX_GPU_DEVICE amrex::Real operator() (int i, int j) const noexcept {
+        using namespace amrex::literals;
+        return (1.0_rt-rel_z)*arr_this(i, j, lo2) + rel_z*arr_prev(i, j, lo2);
+    }
+};
+
+struct interpolated_field_z {
+    // use brace initialization as constructor
+    amrex::MultiFab mfab_this; // field to interpolate on this slice
+    amrex::MultiFab mfab_prev; // field to interpolate on previous slice
     amrex::Real rel_z; // mixing factor between f_view_this and f_view_prev for z interpolation
 
     // use .array(mfi) like with amrex::MultiFab
-    interpolated_field_inner<interp_order_xy> array (amrex::MFIter& mfi) const {
-        amrex::Box bx = f_view_this[mfi].box();
-        return interpolated_field_inner<interp_order_xy>{
-            f_view_this.array(mfi), f_view_prev.array(mfi),
-            1/geom.CellSize(0), 1/geom.CellSize(1),
-            GetPosOffset(0, geom, bx), GetPosOffset(1, geom, bx),
-            rel_z, bx.smallEnd(2)};
+    interpolated_field_z_inner array (amrex::MFIter& mfi) const {
+        return interpolated_field_z_inner{
+            mfab_this.array(mfi), mfab_prev.array(mfi), rel_z, mfab_this[mfi].box().smallEnd(2)};
+    }
+};
+
+template<int interp_order_xy>
+using interpolated_field_xyz = interpolated_field_xy<interp_order_xy, interpolated_field_z>;
+
+struct guarded_field_inner {
+    // captured variables for GPU
+    amrex::Array4<amrex::Real const> array;
+    amrex::Box bx;
+
+    template<class...Args>
+    AMREX_GPU_DEVICE amrex::Real operator() (int i, int j, int k, Args...args) const noexcept {
```

Comment:
I like your style, very sophisticated :)

Alternatively, you can write these as lambdas if you like it very compact:
https://github.com/ECP-WarpX/WarpX/pull/2314

Reply:
Does not work well here because you cannot return a __device__ lambda form a function. Most of these helpers are used in multiple places and I wanted to keep `guarded_field` in the same style.